[excel] (Samples) Punch clock sample (#487)
OfficeDev/office-scripts-docs: docs/resources/scenarios/punch-clock.md

---
title: 'Office Scripts sample scenario: Punch clock button'
description: This sample adds a punch clock button and allows a user to clock in and clock out using the current time.
ms.date: 04/26/2022
ms.localizationpriority: medium
---

# Office Scripts sample scenario: Punch clock button

The scenario idea and script used in this sample was contributed by Office Scripts community member [Brian Gonzalez](https://github.com/b-gonzalez).

> In this scenario, you'll create a time sheet for an employee that allows them to record their start and end times with the press of a [button](../../develop/script-buttons.md). Based on what's previously been recorded, pressing the button will either start their day (clock in) or end their day (clock out). The sample works for both Excel on the web or on Windows.

Reviewer: ```suggestion
In this scenario, you'll create a time sheet for an employee that allows them to record their start and end times with the press of a [button](../../develop/script-buttons.md). Based on what's previously been recorded, pressing the button will either start their day (clock in) or end their day (clock out). The sample works for both Excel on the web and on Windows.
```

          durationString += " and ";
        }
      }
    
      if (hours >= 1) {
        durationString += hours;
        durationString += (hours > 1 ? " hours" : " hour");
        if (minutes >= 1) {
          durationString += " and ";
        }
      }
    
      if (minutes >= 1) {
        durationString += minutes;
        durationString += (minutes > 1 ? " minutes" : " minute");
      }
    
      return durationString;
    }
    ```

1. Rename the script to "Punch clock".

1. Save the script.

1. In the workbook, select cell **E2**.

1. Add a script button. Go to the **More options (…)** menu in the **Script details** page and select **Add button**.

    :::image type="content" source="../../images/punch-clock-sample-2.png" alt-text="The 'More options' menu and the 'Add button' button.":::

1. Save the workbook.

## Run the script

Press the **Punch clock** button to run the script. It either logs the current time under "Clock In" or "Clock Out", depending on what was previously entered.

:::image type="content" source="../../images/punch-clock-sample-3.png" alt-text="The table and the 'Punch clock' button in the workbook.":::

> [!NOTE]

> > The duration is only recorded if it is longer than a minute. Manually edit the "Clock In" time to test larger durations.

Reviewer: ```suggestion
> The duration is only recorded if it's longer than a minute. Manually edit the "Clock In" time to test larger durations.
```

---
title: 'Office Scripts sample scenario: Punch clock button'
description: A sample that lets a user press a button to record the current time.
ms.date: 04/26/2022
ms.localizationpriority: medium

> ---

Reviewer 2: Could we give a shoutout somewhere here to Brian who contributed this script? I liked what we did here with Leslie's sample script: https://docs.microsoft.com/en-us/office/dev/scripts/resources/samples/community-seasons-greetings - perhaps we could link to Brian's script repo here: https://github.com/b-gonzalez/Office-Scripts/

# Office Scripts sample scenario: Punch clock button

> In this scenario, you're setting up a time sheet for an employee. You want to let them record their start and end times with the press of a button, whether they're not Excel for the web or Excel for Desktop.

Reviewer 2: nit: whether they're on Excel for the web or Excel for Desktop

You'll develop a script and a [button](../../develop/script-buttons.md) that records when the user presses the button. Based on what's previously been recorded, pressing the button will either start their day (clock in) or end their day (clock out).

:::image type="content" source="../../images/punch-clock-sample-3.png" alt-text="A table with three columns: 'Clock In', 'Clock Out', and 'Duration' and a button labeled 'Punch clock' in the workbook.":::

## Scripting skills covered

- Script buttons
- Dates and times

## Setup instructions

> 1. Download <a href="punch-clock-sample.xlsx">grade-calculator.xlsx</a> to your OneDrive.

Reviewer 2: This should say "Download punch-clock-sample.xlsx" right?

---

> title: 'Office Scripts sample scenario: Punch clock button'

Reviewer 3: ```suggestion
title: 'Office Scripts sample scenario: Punch time clock button'
```

> description: A sample that lets a user press a button to record the current time.

Reviewer 3: ```suggestion
description: This sample adds a punch clock button and allows a user to log in and log out using the current time.
```

Author: I'll take the spirit of the suggestion, but use "clock" in and out instead of log in, so it doesn't imply an auth path is used.

ms.date: 04/26/2022
ms.localizationpriority: medium
---

# Office Scripts sample scenario: Punch clock button

The scenario idea and script used in this sample was contributed by Office Scripts community member [Brian Gonzalez](https://github.com/b-gonzalez).

> In this scenario, you're setting up a time sheet for an employee. You want to let them record their start and end times with the press of a button, whether they're using Excel on the web or on Windows.

Reviewer 3: ```suggestion
In this scenario, you will create a time sheet for an employee that allows them to record their start and end times with the press of a [button](../../develop/script-buttons.md). Based on what's previously been recorded, pressing the button will either start their day (clock in) or end their day (clock out). The sample works for both Excel on the web or on Windows.
```

> You'll develop a script and a [button](../../develop/script-buttons.md) that records when the user presses the button. Based on what's previously been recorded, pressing the button will either start their day (clock in) or end their day (clock out).

Reviewer 3: ```suggestion

```

:::image type="content" source="../../images/punch-clock-sample-3.png" alt-text="A table with three columns ('Clock In', 'Clock Out', and 'Duration') and a button labeled 'Punch clock' in the workbook.":::

> ## Scripting skills covered

Reviewer 3: ```suggestion
```
This heading and two bullet points seems unnecessary.

> - Script buttons

Reviewer 3: ```suggestion
```

> - Dates and times

Reviewer 3: ```suggestion
```

## Setup instructions

1. Download <a href="punch-clock-sample.xlsx">punch-clock-sample.xlsx</a> to your OneDrive.

    :::image type="content" source="../../images/punch-clock-sample-1.png" alt-text="A table with three columns: 'Clock In', 'Clock Out', and 'Duration'.":::

1. Open the workbook with Excel for the web.

1. Under the **Automate** tab, select **New Script** and paste the following script into the editor.

    ```typescript
    /**
     * This script records either the start or end time of a shift, 
     * depending on what is filled out in the table. 

>      * It is intended to be used with a Script Button.

Author: I know the default is contractions, but I think this reads better in the code comment. Makes the directive clearer by intentionally changing the tone.

     */
    function main(workbook: ExcelScript.Workbook) {
      // Get the first table in the timeSheet.
      const timeSheet = workbook.getWorksheet("MyTimeSheet");
      const timeTable = timeSheet.getTables()[0];
    
      // Get the appropriate table columns.

>       const clockInColumn = timeTable.getColumnByName("Clock in");

Reviewer: ```suggestion
      const clockInColumn = timeTable.getColumnByName("Clock In");
```

     */
    function getDurationMessage(delta: number) {
      // Adapted from here:
      // https://stackoverflow.com/questions/13903897/javascript-return-number-of-days-hours-minutes-seconds-between-two-dates
    
      delta = delta / 1000;
      let durationString = "";
    
      let days = Math.floor(delta / 86400);
      delta -= days * 86400;
    
      let hours = Math.floor(delta / 3600) % 24;
      delta -= hours * 3600;
    
      let minutes = Math.floor(delta / 60) % 60;
    
      if (days >= 1) {
        durationString += days;
        durationString += (days > 1 ? " days" : " day");
    
        if (hours >= 1 && minutes >= 1) {
          durationString += ", ";
        }
        else if (hours >= 1 || minutes > 1) {
          durationString += " and ";
        }
      }
    
      if (hours >= 1) {
        durationString += hours;
        durationString += (hours > 1 ? " hours" : " hour");
        if (minutes >= 1) {
          durationString += " and ";
        }
      }
    
      if (minutes >= 1) {
        durationString += minutes;
        durationString += (minutes > 1 ? " minutes" : " minute");
      }
    
      return durationString;
    }
    ```

> 1. Rename the script to "Punch clock"

Reviewer 3: ```suggestion
1. Rename the script to "Punch clock."
```

Author: Will do, though I'll move the period outside of the quotation marks, so it's clear the user shouldn't put that in the name.

1. Save the script.

1. In the workbook, select cell **E2**.

> 1. Add a script button. Go to the **More options (…)** menu in the **Script details** page and select **Add button***.

Reviewer 3: ```suggestion
1. Add a script button. Go to the **More options (…)** menu in the **Script details** page and select **Add button**.
```

    :::image type="content" source="../../images/punch-clock-sample-2.png" alt-text="The 'More options' menu and the 'Add button' button.":::

1. Save the workbook.

> ## Running the script

Reviewer 3: ```suggestion
## Run the script
```
